src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,93 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = cloneWheels(wheels);
+        if (engine == null) {
+            this.engine = null;
+        } else {
+            this.engine = engine.clone();
+        }
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return cloneWheels(wheels);
+    }
+
+    public Engine getEngine() {
+        if (engine == null) {
+            return null;
+        } else {
+            return engine.clone();
+        }
+    }
+
+    public List<Wheel> cloneWheels(List<Wheel> wheels) {
+        List<Wheel> newListWheels = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> newListWheels = new ArrayList<>(wheels.size());
```
if we know the list size is more than 10, it's good to set it from the beginning